### PR TITLE
UDS: implement GetApplicationData

### LINKED ISSUE #3709: Dragon Ball Fusions - No Working Local Multiplayer
image

The problem with this game is that the communication is only 1 way.
One person can see the other but not the other way around.
Tested on nightly (version in the pic).

[citra_log.txt](https://github.com/citra-emu/citra/files/1966561/citra_log.txt)

### PR DESCRIPTION
**Pull request by mnml**
My goal is mainly to fix multiplayer in Final Fantasy Explorers, which this appears to do successfully (Also fixes #3709).

That being said, I don't really have a good understanding of the 3DS's IPC protocols, so if someone more knowledgeable could check my implementation for correctness, I'd really appreciate it.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5533)

### CODE REVIEW

File: src/core/hle/service/nwm/nwm_uds.h [hunk 1]
```diff
@@ -229,6 +229,18 @@ class NWM_UDS final : public ServiceFramework<NWM_UDS> {
...
     void SetApplicationData(Kernel::HLERequestContext& ctx);
 
+    /**
+     * NWM_UDS::GetApplicationData service function.
+     * Loads the application data from the current beacon.
+     *  Inputs:
+     *      1 : Data size.
+     *  Outputs:
+     *      0 : Return header
+     *      1 : Result of function, 0 on success, otherwise error code
```
Comment (BreadFish64) [inline src/core/hle/service/nwm/nwm_uds.h:239]: It looks like 0 is actually the error code here?
https://www.3dbrew.org/wiki/NWMUDS:GetApplicationData

Comment (mnml) [reply src/core/hle/service/nwm/nwm_uds.h]: Yeah, copy/paste error on my part. Changed it to "always 0".

**Comment by MojoJojoDojo** [comment]
Is there a way I could try this? maybe a test build.
So i could how other games work with this.

**Comment by mnml** [comment]
@MojoJojoDojo You can check the artifacts on AppVeyor: https://ci.appveyor.com/project/bunnei/citra/builds/34939015/job/lgaihhkacypp7dgd/artifacts

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 2]
```diff
@@ -1275,6 +1275,25 @@ void NWM_UDS::SetApplicationData(Kernel::HLERequestContext& ctx) {
...
 
+void NWM_UDS::GetApplicationData(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x11, 1, 0);
+    u32 input_size = rp.Pop<u32>();
+    u8 appdata_size = network_info.application_data_size;
+
+    IPC::RequestBuilder rb = rp.MakeBuilder(2, 2);
+    rb.Push(RESULT_SUCCESS);
+
+    if (input_size < appdata_size) {
```
Comment (FearlessTobi) [inline src/core/hle/service/nwm/nwm_uds.cpp:1286]: "The actual appdata-copying code just verifies that appdata_size is <**=** input_size"
So shouldn't the comparison be `input_size <= appdata_size` here?

Comment (mnml) [reply src/core/hle/service/nwm/nwm_uds.cpp:1286]: Wouldn't that cause the function to return early and not copy the appdata when `input_size == appdata_size`?

Comment (FearlessTobi) [reply src/core/hle/service/nwm/nwm_uds.cpp:1286]: Yes, my bad you're absolutely right.

**Review by B3n30 — APPROVED** [review]

**Pull request merged**